### DISCUSSION TITLE
RFC: com.atproto.sync.getZstdStream - Reducing Relay Egress by 90%

### DISCUSSION THREAD

**Discussion by ybzeek** [bluesky-social/atproto#4582] [Protocol (atproto)]
# RFC: com.atproto.sync.getZstdStream
**Reducing Relay Egress by 90% via Dictionary-Aware Compression and Silicon-Level Optimization.**

**Author:** Zachary Hester  
**Project Reference:** [ybzeek/The-Loom](https://github.com/ybzeek/The-Loom) | [ybzeek/Loom-V2-The-Sovereign-Relay](https://github.com/ybzeek/Loom-V2-The-Sovereign-Relay)  
**Date:** January 2026

---

## 1. Summary
I am proposing a new subscription Lexicon, `com.atproto.sync.getZstdStream`. This endpoint provides a high-efficiency alternative to the current CAR-based firehose by utilizing **Zstandard (zstd) with pre-trained dictionaries**. When paired with a Rust-native processing layer, this architecture reduces relay egress by ~90% and computational overhead by several orders of magnitude, effectively democratizing the role of Relay providers.

## 2. Problem: The "Egress Tax" on Truth
The current firehose implementation (Sync v1/v1.1) broadcasts updates using binary CAR blocks. While secure, this transport layer suffers from:

* **Structural Redundancy:** Repeated Lexicon NSIDs (`app.bsky.feed.post`) and infrastructure strings (`did:plc:`) account for a significant percentage of the stream.
* **The Bandwidth Barrier:** 30–50 Mbps constant egress costs ~$300–$500/mo, creating a centralizing monopoly where only well-funded entities can provide the "Truth" stream.
* **Marshalling Overheads:** High-level middleware (JS/Go) struggles to verify signatures at full network velocity without massive hardware allocation.

## 3. Proposed Solution: The Sovereign Relay
By shifting from "Archive-as-Transport" (CAR) to "Stream-Aware-Transport" (Zstd), we transform the firehose into a lightweight pipe.

### A. Dictionary-Aware Compression
The core breakthrough involves a pre-shared Zstd dictionary trained on global ATmosphere events.

* **Mechanism:** Both client and server load a static dictionary (e.g., `atproto-v1.zdict`). This allows the compression algorithm to replace common strings and Merkle Search Tree (MST) headers with tiny back-references.
* **Impact:** Average event size drops from ~1.2 KB to **~150 bytes**, and specific event sizes for "Likes" drop to **~17 bytes**.

### B. Silicon-Level Processing (Rust)
To mitigate the CPU cost of decompression, I propose a Rust-native implementation (as demonstrated in `Loom-V2`).

* **Zero-Copy Pipelines:** Data is verified and processed in-place, avoiding the "double-copy" memory overhead typical of Node.js or Python.
* **Hardware Acceleration:** Utilizes SIMD (AVX/NEON) for parallel FSE/Huffman decoding, allowing a Raspberry Pi 5 to handle the full global firehose.

## 4. The Lexicon Schema
This Lexicon defines a subscription that expects a `dictId` to ensure the client and relay are using the same "vocabulary."

```json
{
  "lexicon": 1,
  "id": "com.atproto.sync.getZstdStream",
  "defs": {
    "main": {
      "type": "subscription",
      "description": "Subscribe to a network firehose compressed with Zstd + Pre-shared Dictionary.",
      "parameters": {
        "type": "params",
        "properties": {
          "cursor": { "type": "integer", "description": "The sequence number to resume from." },
          "dictId": { "type": "string", "description": "Hash of the Zstd dictionary in use." }
        }
      },
      "message": {
        "type": "union",
        "refs": [ "#frame", "#info" ]
      }
    },
    "frame": {
      "type": "object",
      "required": [ "data", "seq" ],
      "properties": {
        "data": { "type": "bytes", "description": "Zstd-compressed DAG-CBOR payload." },
        "seq": { "type": "integer" }
      }
    },
    "info": {
      "type": "object",
      "properties": {
        "name": { "type": "string" },
        "message": { "type": "string" }
      }
    }
  }
}
```

## 5. Performance Benchmarks
Based on the `Loom-V2` proof-of-concept:

| Metric | Legacy CAR Firehose | Loom-V2 (Zstd-Stream) |
| :--- | :--- | :--- |
| **Monthly Egress** | ~10.5 TB | **~1.1 TB** |
| **Avg. Packet Size** | 1,180 bytes | **145 bytes** |
| **Verification Latency** | ~60ms | **<1.5ms** |
| **Op Throughput (Rust)** | 2k ops/sec | **50k+ ops/sec** |

## 6. Conclusion
The "Big Data" problem is a transport-layer flaw, not a protocol-level one. By hardening our pipes with Rust and Zstandard, we ensure that a **Sovereign Relay** can be run by any individual, returning the "Truth" of the network to its users.

### Project Links & Discussion
The implementation details for this RFC can be found in the project references at the top of this post. 

* **GitHub:** [ybzeek/The-Loom](https://github.com/ybzeek/The-Loom)
* **Status:** Proof-of-concept functional; seeking feedback on Lexicon integration.

I look forward to discussing the technical trade-offs of Zstd dictionary management with the community.

**Comment by notjuliet** [comment]
so you made a shittier [jetstream](https://github.com/bluesky-social/jetstream)

good job Einstein

**Reply by OneDeuxTriSeiGo** [reply]
Wait how so? 

Jetstream doesn't allow you to verify MST signatures though right?

Also pretrained dictionaries seem like they are probably advantageous given the extremely regular structure enforced by the lexicons.

Not to say it's a perfect impl or that it replaces jetstream at all but it seems like a neat experiment for someone to hack together and present (albeit probably better presented in a blog post initially than as an RFC right off the bat. doubly so given that [bluesky-social/atproto](https://github.com/bluesky-social/atproto) is the wrong repo to be submitting RFCs to in the first place as those go to [bluesky-social/proposals](https://github.com/bluesky-social/proposals)).

**Reply by ybzeek** [reply]
Shoot. My bad, I was unsure of where exactly to post something like this. Thank you for informing me though I will make sure to dig harder in the future about proper placement. And yes Jetstream doesn't have any of the crypto verification. It's stripped down to just the jsonl structure and is considered a "trust the provider" pipeline.

**Reply by OneDeuxTriSeiGo** [reply]
yeah no worries. TBH I'd say the best way to share something like this going forward would be to put together a whitewind or leaflet post and share it on bluesky with like the `#atproto` or `#atdev` tags. That way the community can engage with it a bit more organically and not forced to engage with it from a maintainers perspective. Especially since nowadays there are multiple relay implementations, etc so just sharing here probably won't reach all the people interested.

**Reply by OneDeuxTriSeiGo** [reply]
Then if it's well received that'd be the greenlight to submit an "official" RFC that maintainers can engage with with their maintainer hats on.

**Reply by ybzeek** [reply]
Thank you. I will do this in the future.

**Reply by notjuliet** [reply]
I will point out the implementation linked seems in no way to be connecting to the PDSes directly, not even an existing relay, instead it... connects to jetstream directly? which means all the "verification" it claims to perform on top is pointless

Nothing here proves it does any actual crypto verification of the original DAG-CBOR encoded event

**Reply by ybzeek** [reply]
Thanks for having a look at the code. I appreciate the scrutiny, but there are a few technical inaccuracies in your assessment that I'd like to clarify.

1. "Connects to Jetstream directly" -> Incorrect.
The implementation does not connect to Jetstream. As seen in beacon.rs, the default endpoint is strictly wss://bsky.network/xrpc/com.atproto.sync.subscribeRepos (the Firehose).
Furthermore, the parser in lib.rs is a zero-copy DAG-CBOR slicer. Jetstream emits JSON. If this tool were connected to Jetstream, the binary CBOR parser would fail on the first byte.

2. "Verification is pointless / No actual crypto verification" -> Nuanced.
The project explicitly defines itself as a "Mechanical Sympathy" experiment to benchmark the computational limits of consumer hardware (Ryzen 3600).

You are correct that it does not currently perform live DID PLc lookups (PKI resolution). However, it does perform the actual AVX2 cryptographic verification ed25519_dalek::VerifyingKey::verify.

In Live Mode: It parses real DAG-CBOR traffic to test ingestion/bandwidth limits.
In Flood Mode (Benchmark): It verifies cryptographically valid signatures generated by flood_test.rs against a static key.
The goal was to prove that a single CPU has the cycle budget to verify 600k+ Ed25519 signatures per second. The math is real, the CPU load is real, and the AVX2 utilization is real. We are benchmarking the silicon, not the network's identity directory.

If you look at this flamegraph(the far right side) you can see the cpu does spend most of its time inside of ed25519_dalek library performing vector math.
![flamegraph](https://github.com/user-attachments/assets/99ae1d64-0c85-42f2-868f-e65df3ec5ba4)

**Reply by DavidBuchanan314** [reply]
atproto does not (currently) support ed25519, so whatever you're verifying in there doesn't have anything to do with atproto

**Reply by ybzeek** [reply]
I used ed25519 here specifically to benchmark the architecture's memory model against a high-performance AVX2 target.

Switching to k256 would definitively lower the TPS (likely by ~50-60%) as the math is computationally heavier. The core thesis stands, however: the zero-copy plumbing ensures the system remains CPU-bound on the verification rather than I/O-bound.

I'm happy to swap the crypto backend and run a representational k256 benchmark if you'd like to see the exact numbers for that curve.

**Reply by OneDeuxTriSeiGo** [reply]
>     1. "Connects to Jetstream directly" -> Incorrect.
>        The implementation does not connect to Jetstream. As seen in beacon.rs, the default endpoint is strictly wss://bsky.network/xrpc/com.atproto.sync.subscribeRepos (the Firehose).
>        Furthermore, the parser in lib.rs is a zero-copy DAG-CBOR slicer. Jetstream emits JSON. If this tool were connected to Jetstream, the binary CBOR parser would fail on the first byte.

Oh I see. You link the V2 version of the repo at the top but not at the bottom with the project link. The V2 does connect to `wss://bsky.network/xrpc/com.atproto.sync.subscribeRepos` but the V1 seems to connect to jetstream instead.

**Reply by ybzeek** [reply]
Yes the "Loom-V2" and "The Loom" are 2 different repos that do fundamentally different things. The goal was to eventually combine the 2 repos into a singular relay project. But they do work off of the same fundamental architecture.

I apologize if I was not super clear on the documentation or the goal of the RFC.

**Reply by ybzeek** [reply]
I've swapped the backend to [k256] to match the spec.

Updated Results:

Throughput: ~70,000 verified events/sec.
The [k256] math is significantly heavier (approx 9x penalty), but the architecture holds up. We are still CPU-bound on verification, and ~70k TPS is still comfortably above current network volume.

I apologize for not using the correct crypto when I originally released the repo. That's a serious misstep on my end.
